eval/topdown_test.go

```diff
@@ -0,0 +1,540 @@
+		}
+	}
+}
+
+func TestPlugValue(t *testing.T) {
+
+	a := opalog.Var("a")
+	b := opalog.Var("b")
+	c := opalog.Var("c")
+	k := opalog.Var("k")
+	v := opalog.Var("v")
+	cs := parseTerm("[c]").Value
+	ks := parseTerm(`{k: "world"}`).Value
+	vs := parseTerm(`{"hello": v}`).Value
+	hello := opalog.String("hello")
+	world := opalog.String("world")
+
+	ctx1 := &TopDownContext{Bindings: make(Bindings)}
+	ctx1 = ctx1.Bind(a, b)
+	ctx1 = ctx1.Bind(b, cs)
+	ctx1 = ctx1.Bind(c, ks)
+	ctx1 = ctx1.Bind(k, hello)
+
+	ctx2 := &TopDownContext{Bindings: make(Bindings)}
+	ctx2 = ctx2.Bind(a, b)
+	ctx2 = ctx2.Bind(b, cs)
+	ctx2 = ctx2.Bind(c, vs)
+	ctx2 = ctx2.Bind(v, world)
+
+	expected := parseTerm(`[{"hello": "world"}]`).Value
+
+	r1 := plugValue(a, ctx1.Bindings)
+
+	if !expected.Equal(r1) {
+		t.Errorf("Expected %v but got %v", expected, r1)
+		return
+	}
+
+	r2 := plugValue(a, ctx2.Bindings)
+
+	if !expected.Equal(r2) {
+		t.Errorf("Expected %v but got %v", expected, r2)
+	}
+}
+
+func TestTopDownScalarDoc(t *testing.T) {
+	tests := []struct {
+		note     string
+		rule     string
+		expected string
+	}{
+		{"undefined", "p = null :- false", ""}, // "" will be converted to Undefined
+		{"null", "p = null :- true", "null"},
+		{"bool: true", "p = true :- true", "true"},
+		{"bool: false", "p = false :- true", "false"},
+		{"number: 3", "p = 3 :- true", "3"},
+		{"number: 3.0", "p = 3.0 :- true", "3.0"},
+		{"number: 66.66667", "p = 66.66667 :- true", "66.66667"},
+		{`string: "hello"`, `p = "hello" :- true`, `"hello"`},
+		{`string: ""`, `p = "" :- true`, `""`},
+		{"array: [1,2,3,4]", "p = [1,2,3,4] :- true", "[1,2,3,4]"},
+		{"array: []", "p = [] :- true", "[]"},
+		{`object/nested composites: {"a": [1], "b": [2], "c": [3]}`,
+			`p = {"a": [1], "b": [2], "c": [3]} :- true`,
+			`{"a": [1], "b": [2], "c": [3]}`},
+	}
+
+	data := loadSmallTestData()
+
+	for i, tc := range tests {
+
+		ctx := &TopDownContext{
+			Rule:     parseRule(tc.rule),
+			Store:    NewStorageFromJSONObject(data),
+			Bindings: make(map[opalog.Var]opalog.Value),
+		}
+
+		expected := loadExpectedResult(tc.expected)
+		result, err := TopDownQuery(ctx)
+
+		if err != nil {
+			t.Errorf("Test case %d (%v): unexpected error: %v", i+1, tc.note, err)
+			continue
+		}
+
+		if !reflect.DeepEqual(result, expected) {
+			t.Errorf("Test case %d (%v): expected %v but got: %v", i+1, tc.note, expected, result)
+		}
+	}
+
+}
+
+func TestTopDownSetDoc(t *testing.T) {
+
+	tests := []struct {
+		note     string
+		rule     string
+		expected string
+	}{
+		{"array values", "p[x] :- a[i] = x", `[1, 2, 3, 4]`},
+		{"array indices", "p[x] :- a[x] = _", `[0, 1, 2, 3]`},
+		{"object keys", "p[x] :- b[x] = _", `["v1", "v2"]`},
+		{"object values", "p[x] :- b[i] = x", `["hello", "goodbye"]`},
+		{"nested composites", "p[x] :- f[i] = x", `[{"xs": [1.0], "ys": [2.0]}, {"xs": [2.0], "ys": [3.0]}]`},
+		{"deep ref/heterogeneous", "p[x] :- c[i][j][k] = x", `[null, 3.14159, true, false, true, false, "foo"]`},
+	}
```

Comment:
How about one where we're constructing a composite in the body?  Maybe..
p[x] :- x = [i, a[i]]